### PR TITLE
Add methods to Builder object

### PR DESCRIPTION
**Pull request by GuillaumeGomez**

### CODE REVIEW

File: src/builder.rs [hunk 1]
```diff
@@ -57,4 +57,31 @@ impl Builder {
                 .and_then(|obj| obj.downcast().ok())
         }
     }
+
+    pub fn add_from_file(&self, file_name: &str) -> bool {
```
Comment (EPashkin) [inline src/builder.rs]: Why not `&Path`?

Comment (EPashkin) [inline src/builder.rs]: `new_from_file` IMHO need `&Path` also (or some `T: AsPath`)

Comment (GuillaumeGomez) [inline src/builder.rs]: Indeed! Well saw!

**Comment by GuillaumeGomez** [comment]
Changed files path to `&Path`.

**Comment by EPashkin** [comment]
IMHO to more generic `AsPath` better but :+1: anyway.
`xxx_resource` also use file path? I though it some internal path.

File: src/builder.rs [hunk 2]
```diff
@@ -26,13 +28,13 @@ impl Builder {
     }
 
     #[cfg(feature = "v3_10")]
-    pub fn new_from_file(file_name: &str) -> Builder {
+    pub fn new_from_file(file_path: &Path) -> Builder {
```
Comment (gkoz) [inline src/builder.rs]: Elsewhere we use `T: AsRef<Path>`.

File: src/builder.rs [hunk 3]
```diff
@@ -26,13 +28,13 @@ impl Builder {
...
         assert_initialized_main_thread!();
-        unsafe { from_glib_full(ffi::gtk_builder_new_from_file(file_name.to_glib_none().0)) }
+        unsafe { from_glib_full(ffi::gtk_builder_new_from_file(file_path.to_glib_none().0)) }
     }
 
     #[cfg(feature = "v3_10")]
-    pub fn new_from_resource(resource_path: &str) -> Builder {
+    pub fn new_from_resource(resource_path: &Path) -> Builder {
```
Comment (gkoz) [inline src/builder.rs]: The resource paths should probably remain as strings, I don't think resources can handle arbitrary encodings.

File: src/builder.rs [hunk 4]
```diff
@@ -57,4 +59,31 @@ impl Builder {
                 .and_then(|obj| obj.downcast().ok())
         }
     }
+
+    pub fn add_from_file(&self, file_path: &Path) -> bool {
```
Comment (gkoz) [inline src/builder.rs]: We have means to implement proper error handling now.

Comment (GuillaumeGomez) [inline src/builder.rs]: We do?! :open_mouth:

File: src/builder.rs (hunk 3 above)
Comment (GuillaumeGomez) [inline src/builder.rs]: I don't see why it couldn't. The description says it's a path to a resource file after all. What would be the difference with another file path?

Comment (gkoz) [inline src/builder.rs]: It's a path inside the compiled resource bundle not in the file system.

File: src/builder.rs (hunk 4 above)
Comment (gkoz) [inline src/builder.rs]: Grep the generated code for `Error`.

**Comment by EPashkin** [comment]
May be simpler just generate `Builder` and copy paste from it (if not complete generated)?

File: src/builder.rs (hunk 3 above)
Comment (GuillaumeGomez) [inline src/builder.rs]: Ooooooh. Totally misunderstood. Then `Path` isn't appropriate at all!

**Comment by EPashkin** [comment]
From https://developer.gnome.org/gtk3/stable/GtkBuilder.html no errors in it. On any error program just aborted :cry:
Also generating not work, as no `filename` type for `GtkBuilder` functions in .gir.

**Comment by GuillaumeGomez** [comment]
> May be simpler just generate Builder and copy paste from it (if not complete generated)?

Not sure since we provide a very little set of functions.

> From https://developer.gnome.org/gtk3/stable/GtkBuilder.html no errors in it. On any error program just aborted :cry:
> Also generating not work, as no filename type for GtkBuilder functions in .gir.

Arf. That's problematic.

**Comment by gkoz** [comment]
> From https://developer.gnome.org/gtk3/stable/GtkBuilder.html no errors in it. On any error program just aborted :cry:

The way I see it, this applies to constructors (because constructors don't fail) but the `add_*` methods do return errors.

**Comment by EPashkin** [comment]
My bad, I looked only on constructor and missed errors in functions.

File: src/builder.rs [hunk 5]
```diff
@@ -57,4 +61,34 @@ impl Builder {
                 .and_then(|obj| obj.downcast().ok())
         }
     }
+
+    pub fn add_from_file<T: AsRef<Path>>(&self, file_path: T) -> Result<bool, Error> {
```
Comment (EPashkin) [inline src/builder.rs]: IMHO `Result<(), Error>` here and all others

File: src/builder.rs [hunk 6]
```diff
@@ -57,4 +61,34 @@ impl Builder {
...
+        unsafe {
+            let mut error = ::std::ptr::null_mut();
+            let ret = ffi::gtk_builder_add_from_file(self.to_glib_none().0,
```
Comment (EPashkin) [inline src/builder.rs]: `let _ = ` without `> 0`

File: gir [hunk 7]
```diff
@@ -1 +1 @@
-Subproject commit e6cb5d03d8c165d41e93063ef4f59a3559b76169
+Subproject commit a3f05e3ff6b6917e9ee9e7e0ca7ae3d4da1999fc
```
Comment (gkoz) [inline gir]: This seems unrelated.

Comment (GuillaumeGomez) [inline gir]: Oh come on! This is driving me crazy! :rage:

File: src/builder.rs (hunk 5 above)
Comment (GuillaumeGomez) [inline src/builder.rs]: As I understood it, it could fail without returning an error.

File: src/builder.rs (hunk 6 above)
Comment (gkoz) [inline src/builder.rs]: Or simply ignore the return value, there won't be a warning.

File: src/builder.rs (hunk 5 above)
Comment (EPashkin) [inline src/builder.rs]: I was wrong about failing, if ffi function accept error we must return it.
But we return unit `()` in good case instead `bool`.

**Comment by GuillaumeGomez** [comment]
Updated.

**Comment by EPashkin** [comment]
AppVeyor build succeeded, thanks @GuillaumeGomez

**Comment by gkoz** [comment]
Thanks!
